zkevm-circuits/src/state_circuit.rs

```diff
@@ -1,4 +1,273 @@
+use gadgets::is_zero::{IsZeroChip, IsZeroConfig, IsZeroInstruction};
+use halo2_proofs::{
+    circuit::{Layouter, SimpleFloorPlanner},
+    plonk::{
+        Advice, Circuit, Column, ConstraintSystem, Error, Expression, Fixed, Instance, VirtualCells,
+    },
+    poly::Rotation,
+};
+use lexicographic_ordering::{
+    Chip as LexicographicOrderingChip, Config as LexicographicOrderingConfig,
+};
+use lookups::{Chip as LookupsChip, Config as LookupsConfig, Queries as LookupsQueries};
+use multiple_precision_integer::{Chip as MpiChip, Config as MpiConfig, Queries as MpiQueries};
+use random_linear_combination::{Chip as RlcChip, Config as RlcConfig, Queries as RlcQueries};
+
+const N_LIMBS_RW_COUNTER: usize = 2;
+const N_LIMBS_ACCOUNT_ADDRESS: usize = 10;
+const N_LIMBS_ID: usize = 2;
+
+/// Config for StateCircuit
+#[derive(Clone)]
+pub struct StateConfig<F: Field> {
+    selector: Column<Fixed>, // Figure out why you get errors when this is Selector.
+    // https://github.com/appliedzkp/zkevm-circuits/issues/407
+    rw_counter: MpiConfig<u32, N_LIMBS_RW_COUNTER>,
+    is_write: Column<Advice>,
+    tag: Column<Advice>,
+    id: MpiConfig<u32, N_LIMBS_ID>,
+    address: MpiConfig<Address, N_LIMBS_ACCOUNT_ADDRESS>,
+    field_tag: Column<Advice>,
+    // Consider using WordConfig instead?
+    storage_key: RlcConfig<N_BYTES_WORD>,
+    is_storage_key_unchanged: IsZeroConfig<F>,
+    value: Column<Advice>,
+    lookups: LookupsConfig,
+    power_of_randomness: [Column<Instance>; N_BYTES_WORD - 1],
+    lexicographic_ordering: LexicographicOrderingConfig<F>,
+}
+
+type Lookup<F> = (&'static str, Expression<F>, Expression<F>);
+
+/// State Circuit for proving RwTable is valid
+#[derive(Default)]
+pub struct StateCircuit<F: Field> {
+    pub(crate) randomness: F,
+    pub(crate) rows: Vec<Rw>,
+}
+
+impl<F: Field> StateCircuit<F> {
+    /// make a new state circuit
+    pub fn new(randomness: F, rw_map: RwMap) -> Self {
+        let mut rows: Vec<_> = rw_map.0.into_values().flatten().collect();
+        rows.sort_by_key(|row| {
+            (
+                row.tag() as u64,
+                row.field_tag().unwrap_or_default(),
+                row.id().unwrap_or_default(),
+                row.address().unwrap_or_default(),
+                row.storage_key().unwrap_or_default(),
+                row.rw_counter(),
+            )
+        });
+        Self { randomness, rows }
+    }
+
+    /// powers of randomness for instance columns
+    pub fn instance(&self) -> Vec<Vec<F>> {
+        (1..32)
+            .map(|exp| vec![self.randomness.pow(&[exp, 0, 0, 0]); self.rows.len()])
+            .collect()
+    }
+}
+
+impl<F: Field> Circuit<F> for StateCircuit<F> {
+    type Config = StateConfig<F>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+        let selector = meta.fixed_column();
```

Comment:
what about change selector into advice column?   Since it is not "fixed" now. ( We assign "1" to it only when a row is  non empty, so this column is dynamic.

Reply:
There needs to be a fixed column or selector involved in each of the constraints, or the MockProver will return a `ConstraintPoisoned` error when it tries to verify. 

We can fill up all the otherwise empty rows with `Rw::Start`'s so that the `selector` column will not be dynamic. I haven't quite figured out how to do this without massively slowing down the tests though.

Reply:
verify_at_rows can be helpful for this problem.

https://github.com/scroll-tech/zkevm-circuits/pull/140/commits/34eb52fb2244ab602c04cab0b048ec8f8a9c9644

Reply:
"fill up all the otherwise empty rows with Rw::Start" is also a bit counterintuitive..
Maybe we can merge this PR first(since it may be not solved very cleanly now?), then revisit this problem (make sure vk generated from one circuit instance can verify another circuit) later.